[Custom] Support build command with env variables

dgageot opened the pull request:
Fixes #3082

briandealwis commented on pkg/skaffold/build/custom/custom.go:
```diff
@@ -68,9 +68,15 @@ func (b *ArtifactBuilder) Build(ctx context.Context, out io.Writer, a *latest.Ar
+	var cmd *exec.Cmd
 	split := strings.Split(artifact.BuildCommand, " ")
+	if len(split) > 1 {
+		cmd = exec.Command("sh", "-c", artifact.BuildCommand)
```
This won't work on Windows.

nkubala commented on pkg/skaffold/build/custom/custom.go:
```diff
@@ -68,9 +69,19 @@ func (b *ArtifactBuilder) Build(ctx context.Context, out io.Writer, a *latest.Ar
+
+	var cmd *exec.Cmd
 	split := strings.Split(artifact.BuildCommand, " ")
+	if len(split) > 1 {
```
if we're going to check the platform here anyway, then is there any point in checking `if len(split) > 1`? functionally `exec.Command(artifact.BuildCommand)` and `exec.Command("sh", "-c", artifact.BuildCommand)` should be the same thing right?

dgageot replied:
It's not the same. I'll had a comment

nkubala commented on pkg/skaffold/build/custom/custom.go:
```diff
@@ -68,9 +69,19 @@ func (b *ArtifactBuilder) Build(ctx context.Context, out io.Writer, a *latest.Ar
+	var cmd *exec.Cmd
 	split := strings.Split(artifact.BuildCommand, " ")
+	if len(split) > 1 {
+		if runtime.GOOS == "windows" {
```
also, it feels a little weird to have this check live right in the custom builder implementation. can this maybe be a function `GetShellCommand()` or something in `pkg/skaffold/util/cmd.go`?

dgageot replied:
I didn't want to make things too complicated. I'll what I can do.

dgageot commented:
@briandealwis @nkubala Thanks for the review. I fixed the code for windows and added a comment that explains why we need to handle differently commands with or without args.

briandealwis commented on pkg/skaffold/build/custom/custom.go:
```diff
@@ -68,9 +69,26 @@ func (b *ArtifactBuilder) Build(ctx context.Context, out io.Writer, a *latest.Ar
+		cmd = exec.Command(artifact.BuildCommand)
+
+	// When the user specifies a command with args, such as `./build.sh -flag $IMAGE`,
+	// Skaffold has to shell out to sh in order for arguments to be evaluated.
```
It's possible the executable is an environment variable too.  I define `$JAVA7`, `$JAVA8`, and `$JAVA11` in my environment to invoke the corresponding `java` binary.  Maybe we should just use `sh`/`cmd` always?

dgageot replied:
we could do that

balopat replied:
I'm up for that. This split based determination is a "poor man's parsing" anyway - a proxy for "does this command have env vars"

balopat commented on pkg/skaffold/build/custom/custom.go:
```diff
@@ -68,9 +69,26 @@ func (b *ArtifactBuilder) Build(ctx context.Context, out io.Writer, a *latest.Ar
+
+	// When the user specifies a command with args, such as `./build.sh -flag $IMAGE`,
+	// Skaffold has to shell out to sh in order for arguments to be evaluated.
+	case runtime.GOOS != "windows":
```
final nit: 

can we swap this? 
i.e. `default` would be `sh` and `runtime.GOOS == "windows"` would be `cmd.exe`
